### PR TITLE
Initial design

### PR DESCRIPTION
**Pull request by svandenhoek**
Initial design with implemented LIRICAL benchmark runner. Be sure to check both READMEs (`README.md` & `biobesu/suite/hpo_generank/README.md `) for information on how to run (tests).

### CODE REVIEW

File: biobesu/helper/converters.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        :type phenotype_ids: list[str]
+        :return: a JSON-formatted phenopacket string
+        :rtype: str
+        """
+
+        # Writes opening bracket.
+        output_string = '{'
+
+        # Convert row id to phenopackets.
+        output_string += '\n\t"id": "' + phenopacket_id + '",' + \
```
Comment (marikaris) [inline biobesu/helper/converters.py]: I always prefer using the .format function. Especially when my string spreads across several lines:
```python
output_string += '\n\t"id": "{}",\n\t"phenotypic_features": ['.format(phenopacket_id)
```

File: biobesu/helper/converters.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+                         '\n\t"phenotypic_features": ['
+
+        # Converts phenotypes phenopackets.
+        for i, phenotype_id in enumerate(phenotype_ids):
+            output_string += '{' \
```
Comment (marikaris) [inline biobesu/helper/converters.py]: I would suggest using .format for this one as well

File: biobesu/helper/converters.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+                             '\n\t\t}' \
+                             '\n\t}'
+            if i < len(phenotype_ids)-1:
+                output_string += ', '
+
+        # Closing bracket for phenotypicFeatures.
+        output_string += '],'
+
+        # Add metadata.
+        output_string += '\n\t"meta_data": {' \
```
Comment (marikaris) [inline biobesu/helper/converters.py]: .format

File: biobesu/helper/converters.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    # Download URL (ordered by gene ID).
+    download_file = 'https://www.genenames.org/cgi-bin/download/custom?col=gd_pub_eg_id&col=gd_app_sym' \
+                    '&status=Approved&status=Entry%20Withdrawn&hgnc_dbtag=on&order_by=gd_pub_eg_id' \
+                    '&format=text&submit=submit'
+
+    # Expected file name.
+    file_name = 'gene_ids_symbols.tsv'
+
+    # Expected header format.
+    expected_file_header = "NCBI Gene ID\tApproved symbol"
```
Comment (marikaris) [inline biobesu/helper/converters.py]: Single quotes (because you use them everywhere)

File: biobesu/cli.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+from biobesu.helper.argument_parser import BiobesuParser
+
+
+def main():
+    # Load available suites.
+    suites = retrieve_entry_point('biobesu_suites')
+
+    # Defines global command line.
+    parser = BiobesuParser(formatter_class=RawTextHelpFormatter, add_help=False)
+    parser.add_argument("suite", help="the chosen benchmark suite:\n" + '\n'.join(suites))
```
Comment (marikaris) [inline biobesu/cli.py]: You should use single/double quotes consistently. I'd suggest using single quotes. 

Instead of \n use `os.linesep` for lines written to the commandline (for our beloved windows users)

File: biobesu/helper/converters.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+    Converter for phenotype data.
+    """
+
+    def __init__(self, hpo_obo):
+        # Defines dictionaries for fast retrieval.
+        self.names_by_id = {}
+        self.id_by_names = {}
+
+        # Stores hpo obo relevant information.
+        self.hpo_obo_version = ""
```
Comment (marikaris) [inline biobesu/helper/converters.py]: single quotes

File: biobesu/helper/converters.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+    def __read_file(self):
+        """
+        Digests the conversion file.
+        """
+
+        # Goes through the file.
+        for counter, line in enumerate(open(self.gene_file)):
+            # Validates if all expected columns are present and in expected order.
+            if counter == 0 and line.rstrip() != self.expected_file_header:
+                raise Exception("Unexpected gene info file header.\nExpected: {0}\nActual: {1}"
```
Comment (marikaris) [inline biobesu/helper/converters.py]: Use more specific exception, I'd suggest either a `ValueError` or a custom exception. + use single quotes

File: biobesu/helper/converters.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+                                .format(self.expected_file_header, line))
+
+            # Processes items.
+            if counter > 0:
+                line = line.rstrip().split("\t")
```
Comment (marikaris) [inline biobesu/helper/converters.py]: single quotes

File: biobesu/helper/converters.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+                    line.append('')
+
+                # Clearer usage of data.
+                gene_id = line[0]
+                gene_symbol = line[1]
+
+                # Adds gene to dict.
+                if gene_id != '':
+                    if self.id_by_symbol.get(gene_symbol) is not None:
+                        raise Exception("The symbol {} was already assigned to {}".format(gene_symbol, gene_id))
```
Comment (marikaris) [inline biobesu/helper/converters.py]: More specific error + single quotes

File: biobesu/helper/validate.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+def program(cmd):
+    """
+    Checks whether the command line (which should contain an application name) refers to a valid path. If it is an
+    existing application, does nothing. If the path is not valid, raises an OSError.
+    :param cmd: the name of the application as how it would be called using the command line
+    """
+
+    if not which(cmd):
+        raise OSError("Error: " + cmd + " is not available on this system.")
```
Comment (marikaris) [inline biobesu/helper/validate.py]: single quotes
I prefer .format over the string concatenation (the same goes for the following lines), but here it's not that horrible because it's still quite readable. 
Alternatively you could take a look at [literal string interpolation](https://www.python.org/dev/peps/pep-0498/), but that only works from python 3.6 and up, so maybe that would be less ideal. It is the most readable and efficient method though.

File: biobesu/suite/hpo_generank/cli.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+from biobesu.helper.generic import retrieve_entry_point
+
+
+def main(parser):
+    # Load available runners.
+    runners = retrieve_entry_point('biobesu_hpo_generank')
+
+    # Adds suite-specific command line.
+    parser.add_argument("runner", help="a runner from this suite:\n" + '\n'.join(runners))
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/cli.py]: single quotes + os.linesep

File: biobesu/suite/hpo_generank/runner/lirical.py [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+        # Generate phenopackets.
+        phenopackets_dir = __generate_phenopacket_files(args)
+        # Run lirical.
+        lirical_output_dir = __run_lirical(args, phenopackets_dir)
+        # Extract relevant fields from lirical output.
+        lirical_gene_alias_file, lirical_omims_file = __extract_from_lirical_output(args, lirical_output_dir)
+        # Convert output to genes.
+        __convert_lirical_extractions(args, lirical_gene_alias_file, lirical_omims_file)
+    except FileExistsError as e:
+        print("\nAn output file/directory already exists: " + e.filename + "\nExiting...")
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/runner/lirical.py]: * single quotes
* os.linesep
* possibly .format

File: biobesu/suite/hpo_generank/runner/lirical.py [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+    Parsers the command line
+
+    :param parser: the argument parser
+    :type parser: ArgumentParser
+    :return: the parsed arguments
+    :rtype:
+    """
+
+    # Adds runner-specific command line.
+    parser.add_argument("--jar", required=True, help="location of LIRICAL .jar file")
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/runner/lirical.py]: single quotes (applies to whole `def`)

File: biobesu/suite/hpo_generank/runner/lirical.py [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+def __generate_phenopacket_files(args):
+    """
+    Generates the phenopacket files from the benchmark data.
+
+    :param args: the parsed arguments
+    :return: the directory containing the phenopacket files
+    :rtype: str
+    """
+
+    phenopackets_dir = create_dir(args.output + "phenopackets/")
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/runner/lirical.py]: single quotes

File: biobesu/suite/hpo_generank/runner/lirical.py [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+    Runs lirical for each phenopacket file.
+
+    :param args: the parsed arguments
+    :param phenopackets_dir: the directory containing the phenopacket files
+    :param phenopackets_dir: str
+    :return: the directory containing the LIRICAL output
+    :rtype: str
+    """
+
+    lirical_output_dir = create_dir(args.output + "lirical_output/")
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/runner/lirical.py]: single quotes

File: biobesu/suite/hpo_generank/runner/lirical.py [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+
+    # Run tool for each input file.
+    for file in listdir(phenopackets_dir):
+        file_path = phenopackets_dir + file
+        file_id = file.rstrip(".json").split('/')[-1]
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/runner/lirical.py]: single quotes

File: biobesu/suite/hpo_generank/runner/lirical.py [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+        call("java -jar " + args.jar + " phenopacket -p " + file_path + " -o " + lirical_output_dir + " -x " + file_id +
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/runner/lirical.py]: single quotes + .format/literal string interpolation (if >python 3.6)

File: biobesu/suite/hpo_generank/runner/lirical.py [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+    Extracts the relevant information from the LIRICAL output.
+
+    :param args: the parsed arguments
+    :param lirical_output_dir: the directory containing the LIRICAL output
+    :type lirical_output_dir: str
+    :return: 2 file paths, one to the gene alias file and one to the omim file
+    :rtype: tuple[str,str]
+    """
+
+    extract_dir = create_dir(args.output + "lirical_extraction/")
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/runner/lirical.py]: single quotes (applies to whole `def`)

File: biobesu/suite/hpo_generank/runner/lirical.py [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+    Converts the LIRICAL extracts to a more usable format.
+
+    :param args: the parsed arguments
+    :param lirical_gene_alias_file: the path to the file containing the extracted gene aliases
+    :type lirical_gene_alias_file: str
+    :param lirical_omims_file: the path to the file containing the extracted gene aliases
+    :type lirical_omims_file: str
+    """
+
+    conversion_dir = create_dir(args.output + "lirical_conversion/")
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/runner/lirical.py]: single quotes (whole `def`)

File: setup.py [hunk 20]
```diff
@@ -0,0 +1,46 @@
+#!/user/bin/env python3
+
+from setuptools import setup, find_namespace_packages
+
+with open("README.md", "r", encoding="utf-8") as fh:
```
Comment (marikaris) [inline setup.py]: single quotes (whole `def`)

File: test/helper/converters_test.py [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+def: "Diverticulum (sac or pouch) in the wall of the urinary bladder." [HPO:probinson]
+synonym: "Bladder diverticula" EXACT [HPO:skoehler]
+xref: MSH:C562406
+xref: SNOMEDCT_US:197866008
+xref: UMLS:C0156273
+is_a: HP:0025487 ! Abnormality of bladder morphology
+
+"""
+
+    @patch("builtins.open", new_callable=mock_open, read_data=hpo_obo)
```
Comment (marikaris) [inline test/helper/converters_test.py]: single quotes (whole `class`)

File: biobesu/suite/hpo_generank/runner/lirical.py [hunk 22]
```diff
@@ -0,0 +1,274 @@
...
+            # Process input files.
+            for file in listdir(lirical_output_dir):
+                # Generate ID column.
+                id_column = "\n{}\t".format(file.rstrip(".tsv").split('/')[-1])
+                alias_writer.write(id_column)
+                omim_writer.write(id_column)
+
+                # Create/write genes column.
+                with open(lirical_output_dir + file) as input_file:
+                    genes, omims = __extract_fields_from_lirical_data(input_file)
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/runner/lirical.py:159]: PyCharm says this: 
Screenshot 2021-03-15 at 10 46 23
I'm not sure whether it's relevant, but maybe worth it to take a look.

File: test/helper/converters_test.py [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+        actual_output = converter.id_to_symbol(input_data)
+
+        assert actual_output == expected_output
+
+    @patch("builtins.open", new_callable=mock_open, read_data=gene_info_file)
+    def test_id_to_symbol_list_with_missing(self, mock_open):
+        converter = converters.GeneConverter("/path/to/file")
+
+        input_data = ['10', '3', '14']
+        expected_output = (['NAT2', 'A2MP1'], set(('14',)))
```
Comment (marikaris) [inline test/helper/converters_test.py]: `set(('14',))` can be replaced with `{14}`

File: test/helper/converters_test.py [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+        converter = converters.PhenotypeConverter("/path/to/file")
+
+        hpo_names = ["Abnormality of female internal genitalia", "Bladder diverticulum"]
+        expected_output = (["HP:0000008", "HP:0000015"], set())
+        actual_output = converter.name_to_id(hpo_names)
+
+        assert actual_output == expected_output
+
+    @patch("builtins.open", new_callable=mock_open, read_data=hpo_obo)
+    def test_id_to_phenopackets(selfself, mock_open):
```
Comment (marikaris) [inline test/helper/converters_test.py]: I think one `self` should be enough ;-)

File: test/helper/converters_test.py [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+xref: MSH:C562406
+xref: SNOMEDCT_US:197866008
+xref: UMLS:C0156273
+is_a: HP:0025487 ! Abnormality of bladder morphology
+
+"""
+
+    @patch("builtins.open", new_callable=mock_open, read_data=hpo_obo)
+    def test_id_to_symbol(self, mock_open):
+        converter = converters.PhenotypeConverter("/path/to/file")
```
Comment (marikaris) [inline test/helper/converters_test.py]: Make constant for `'/path/to/file'`, instead of duplicating it throughout this file

File: biobesu/suite/hpo_generank/README.md [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+   --output /path/to/dir/output --lirical_data /path/to/dir/lirical/data \
+   --runner_data /path/to/tmp/dir/
+   ```
+
+Note: `--runner_data` is needed for designating a location where the runner can download temporary data to. When executing the runner multiple times, using the same path skips re-downloading the same data every time.
+
+
+
+
+[lirical_download]: https://github.com/TheJacksonLaboratory/LIRICAL/releases/tag/v1.3.0
```
Comment (marikaris) [inline biobesu/suite/hpo_generank/README.md:32]: dead link

**Review by marikaris — CHANGES_REQUESTED** [review]

File: biobesu/helper/converters.py (hunk 2 above)
Comment (svandenhoek) [reply biobesu/helper/converters.py]: Are we sure we want to use `.format()` for all these?

Any curly bracket used for the json output would need to be escaped:
```
'{{' \
'\n\t\t"type": {{' \
'\n\t\t\t"id": "{}",' \
'\n\t\t\t"label": "{}"' \
'\n\t\t}}' \
'\n\t}}' \
.format()
```

Not sure if this would be that beneficial as it might also make those Strings (a bit) less readable.

Comment (marikaris) [reply biobesu/helper/converters.py]: Right, I agree. For those you might want to keep it this way. I really dislike the `\t\t\n`'s, but I don't see any other way to make it prettier.

Comment (svandenhoek) [reply biobesu/helper/converters.py]: Agreed regarding that it does not look pretty. Did a quick google search and I'll have a look at rewriting the code using [json.dumps()](https://docs.python.org/3/library/json.html#json.dumps)

File: biobesu/helper/converters.py (hunk 1 above)
Comment (svandenhoek) [reply biobesu/helper/converters.py]: See https://github.com/molgenis/biobesu/pull/1#discussion_r594982008

File: biobesu/helper/converters.py (hunk 3 above)
Comment (svandenhoek) [reply biobesu/helper/converters.py]: See https://github.com/molgenis/biobesu/pull/1#discussion_r594982008

File: biobesu/cli.py (hunk 5 above)
Comment (svandenhoek) [reply biobesu/cli.py]: Good point. It seems that argparse uses `\n` itself as well in [its code](https://github.com/python/cpython/blob/3.9/Lib/argparse.py#L230).

I took the liberty of trying it out on my Windows 10 desktop (after installing python3.9 & pip), which resulted in the output in the screenshot when running it through the windows command prompt. I'm not sure why this is formatted correctly (as according to the documentation [you should be right](https://docs.python.org/3/library/os.html#os.linesep)). Perhaps Python already makes this Windows-compatible?
Schermopname (49)

Long story short: I suggest keeping the parts that are input for argparse similar to how argparse handles newlines itself. When we get reports of Windows users with the issue of a wrongly formatted usage, investigate the issue further.

Comment (marikaris) [reply biobesu/cli.py]: Right, that looks okay to me. I commented it because recently I used `\n` in a java application, which caused the application to fail on windows, so I looked for a python alternative, which apparently is `os.linesep`, although you only don't want to use it in text output. I had problems with this before. Especially when running in pycharm rather than directly from the commandline. If both works, I'd suggest using `os.linesep`, just to be sure.

File: biobesu/cli.py [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+
+    # Processes command line.
+    args, unknown_args = parser.parse_known_args()
+
+    # Run selected suite.
+    suites[args.suite](parser)
```
Comment (marikaris) [inline biobesu/cli.py:20]: Please add: 
```python
if __name__ == '__main__':
    main()
```
This allows me to run it in pycharm

File: biobesu/cli.py [hunk 28]
```diff
@@ -0,0 +1,21 @@
+#!/user/bin/env python3
+
+from argparse import ArgumentParser
```
Comment (marikaris) [inline biobesu/cli.py]: According to PyCharm, you don't need this import?

File: biobesu/cli.py (hunk 5 above)
Comment (svandenhoek) [reply biobesu/cli.py]: As `os.linesep` is not needed for files in text-mode, it might treat stdout/stderr the same way. Would explain the output I got on Windows. Furthermore [print()](https://docs.python.org/3/library/functions.html#print) also uses `\n` and not not `os.linesep`. So I would still suggest just keeping `\n` unless we get reports from Windows users regarding missing newlines.

File: biobesu/suite/hpo_generank/cli.py (hunk 11 above)
Comment (svandenhoek) [reply biobesu/suite/hpo_generank/cli.py]: See https://github.com/molgenis/biobesu/pull/1#discussion_r595167899

File: biobesu/suite/hpo_generank/runner/lirical.py (hunk 12 above)
Comment (svandenhoek) [reply biobesu/suite/hpo_generank/runner/lirical.py]: In regards to `os.linesep`: see https://github.com/molgenis/biobesu/pull/1#discussion_r595167899

File: test/helper/converters_test.py (hunk 24 above)
Comment (svandenhoek) [reply test/helper/converters_test.py]: Heh, oops 😅

File: test/helper/converters_test.py (hunk 25 above)
Comment (svandenhoek) [reply test/helper/converters_test.py]: As the actual input is mocked, that String is purely illustrative of what usually is expected as input. Keeping it verbose seems clearer to me than referring to a variable containing a useless String, though I could put empty String if that is preferred?

File: biobesu/suite/hpo_generank/README.md (hunk 26 above)
Comment (svandenhoek) [reply biobesu/suite/hpo_generank/README.md:32]: It seems to work on my side. Could you check again?

```
$ curl --head https://github.com/TheJacksonLaboratory/LIRICAL/releases/tag/v1.3.0
HTTP/2 200
```

File: biobesu/cli.py (hunk 27 above)
Comment (svandenhoek) [reply biobesu/cli.py:20]: Will be added to both `cli.py` scripts & `lirical.py` (might want to validate if they work properly for you as well though before accepting).

File: biobesu/suite/hpo_generank/runner/lirical.py (hunk 22 above)
Comment (svandenhoek) [reply biobesu/suite/hpo_generank/runner/lirical.py:159]: `TextIOWrapper` as type was based on the [open()](https://docs.python.org/3/library/functions.html#open) docs:
> The type of file object returned by the open() function depends on the mode. When open() is used to open a file in a text mode ('w', 'r', 'wt', 'rt', etc.), it returns a subclass of io.TextIOBase (specifically io.TextIOWrapper).

Might be that `TextIO` [is the type](https://docs.python.org/3/library/typing.html#typing.TextIO) whereas `TextIOWrapper` [the implementation](https://docs.python.org/3/library/io.html#io.TextIOWrapper). I'll change it to TextIO.

File: biobesu/suite/hpo_generank/README.md (hunk 26 above)
Comment (svandenhoek) [reply biobesu/suite/hpo_generank/README.md:32]: On second thought, you probably refer to [this typo](https://github.com/molgenis/biobesu/pull/1/files/15a5bc08060ef14899be6a41490a207fc4c13816..67b42d785c9f37dbe58efa2c9842561a673ca5e2#diff-159a76fa3d51da3ee19c226428e7cac7f403e3c8301163edfd77cc572a6e0536L12) which has now been fixed. So in that case this issue can be resolved 🙂

File: test/helper/converters_test.py (hunk 25 above)
Comment (svandenhoek) [reply test/helper/converters_test.py]: Ignore the previous post, replaced the Strings with a descriptive variable name.

File: biobesu/helper/converters.py [hunk 29]
```diff
@@ -0,0 +1,291 @@
...
+        :param phenopacket_id: the ID to be used for the phenopacket
+        :type phenopacket_id str
+        :param phenotype_ids: the phenotypes to convert
+        :type phenotype_ids: list[str]
+        :return: a JSON-formatted phenopacket string
+        :rtype: str
+        """
+
+        # Create json dict with id.
+        json_dict = {"id": phenopacket_id}
```
Comment (marikaris) [inline biobesu/helper/converters.py]: Ha! I found more double quotes (sorry)

File: biobesu/helper/validate.py (hunk 10 above)
Comment (marikaris) [inline biobesu/helper/validate.py]: Still double quotes

File: biobesu/suite/hpo_generank/README.md (hunk 26 above)
Comment (marikaris) [reply biobesu/suite/hpo_generank/README.md:32]: It was that typo, so It's resolved now

File: test/suite/hpo_generank/helper/hpo_generank_converters_test.py [hunk 30]
```diff
@@ -0,0 +1,52 @@
...
+    compress.write(gene_info_string)
+    compress.close()
+
+    file_path_placeholder = ''
+
+    @patch("builtins.open", new_callable=mock_open, read_data=gene_info.getvalue())
+    def test_alias_to_gene_symbol(self, mock_open):
+        converter = converters.LiricalGeneAliasConverter(self.file_path_placeholder)
+
+        aliases = ["GAB", "CPAMD5"]
```
Comment (marikaris) [inline test/suite/hpo_generank/helper/hpo_generank_converters_test.py]: quotes

File: test/suite/hpo_generank/helper/hpo_generank_converters_test.py [hunk 31]
```diff
@@ -0,0 +1,52 @@
...
+100300	57514	phenotype	 GeneMap	C4551482	-
+"""
+
+    file_path_placeholder = ''
+
+    @patch("builtins.open", new_callable=mock_open, read_data=mim2_gene)
+    def test_alias_to_gene_symbol(self, mock_open):
+        converter = converters.LiricalOmimConverter(self.file_path_placeholder)
+
+        omims = ["100300", "100100"]
```
Comment (marikaris) [inline test/suite/hpo_generank/helper/hpo_generank_converters_test.py]: quotes

**Review by marikaris — CHANGES_REQUESTED** [review]
These are more like notes to myself. I'll fix them and then I'll merge it.

**Review by marikaris — APPROVED** [review]
